csharp/Adapter/Microsoft.Spark.CSharp/Core/PriorityQueue.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Spark.CSharp.Core
+{
+    /// <summary>
+    /// A bounded priority queue implemented with max binary heap.
+    /// 
+    /// Construction steps:
+    ///  1. Build a Max Heap of the first k elements.
+    ///  2. For each element after the kth element, compare it with the root of the max heap,
+    ///    a. If the element is less than the root, replace root with this element, heapify.
+    ///    b. Else ignore it.
+    /// </summary>
+    [Serializable]
+    internal class PriorityQueue<T> : IEnumerable<T> where T : IComparable<T>
```

Comment:
@skaarthik this piece of code has already been well covered by current unit tests. Of course I can add some dedicated unit tests for this class in next commit.